File: app/src/main/java/kittentrate/game/GameManager.java
package kittentrate.game;

import java.util.Map;
import java.util.WeakHashMap;

import kittentrate.utils.Constants;

/**
 * Created by Manuel Lorenzo on 21/03/2017.
 */

public class GameManager implements GameDomainContract.Manager {
    private Map<Integer, Card> facingUpCardsWeakHashMap = new WeakHashMap<>(Constants.NUMBER_MATCHING_CARDS);
    private GameContract.Presenter gamePresenter;
    private Game game;

    public GameManager(GameContract.Presenter gamePresenter) {
        this.gamePresenter = gamePresenter;
        this.game = new Game();
    }

    @Override
    public void cardFlipped(int position, Card card) {
        if (!facingUpCardsWeakHashMap.containsKey(position)) {
            facingUpCardsWeakHashMap.put(position, card);
            if (shouldCheckForCardsMatch()) {
                // This would be the last card, we should check for matches.
                String id = card.getId();
                boolean matchFound = true;
                for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
                    Card loopCard = entry.getValue();
                    if (!loopCard.getId().equals(id)) {
                        matchFound = false;
                    }
                }
                if (matchFound) {
                    // Match found; remove cards from dataset
                    gamePresenter.removeViewFlipper();
                    for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
                        Card card1 = entry.getValue();
                        gamePresenter.notifyAdapterItemRemoved(card1.getId());
                    }
                    facingUpCardsWeakHashMap.clear();
                    increaseGameScore();
                    gamePresenter.removeViewFlipper();
                } else {
                    // Match not found; turn cards over and start from the beginning.
                    decreaseGameScore();
                    gamePresenter.onTurnCardsOver();
                }
                gamePresenter.onGameScoreChanged(getGameScore());
            }
        }
    }

    @Override
    public void removeCardsFromMap() {
        for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
            gamePresenter.notifyAdapterItemChanged(entry.getKey());
        }
        facingUpCardsWeakHashMap.clear();
    }

    @Override
    public boolean shouldDispatchUiEvent() {
        return facingUpCardsWeakHashMap.size() != Constants.NUMBER_MATCHING_CARDS;
    }

    @Override
    public void resetScore() {
        resetGameScore();
    }

    private boolean shouldCheckForCardsMatch() {
        return facingUpCardsWeakHashMap.size() == Constants.NUMBER_MATCHING_CARDS;
    }

    private void increaseGameScore() {
        game.setScore(game.getScore() + 1);
    }

    private void decreaseGameScore() {
        game.setScore(game.getScore() - 1);
    }

    private int getGameScore() {
        return game.getScore();
    }

    private void resetGameScore() {
        game.setScore(0);
    }
}

File: app/src/main/java/kittentrate/game/GameManager.kt
package kittentrate.game

import kittentrate.utils.Constants
import java.util.*

/**
 * Created by Manuel Lorenzo on 21/03/2017.
 */

class GameManager(private val gamePresenter: GameContract.Presenter) : GameDomainContract.Manager {
    private val facingUpCardsWeakHashMap = WeakHashMap<Int, Card>(Constants.NUMBER_MATCHING_CARDS)
    private val game: Game = Game()

    private val gameScore: Int
        get() = game.score

    override fun cardFlipped(position: Int, card: Card) {
        if (!facingUpCardsWeakHashMap.containsKey(position)) {
            facingUpCardsWeakHashMap.put(position, card)
            if (shouldCheckForCardsMatch()) {
                // This would be the last card, we should check for matches.
                val id = card.id
                var matchFound = true
                for ((_, loopCard) in facingUpCardsWeakHashMap) {
                    if (loopCard.id != id) {
                        matchFound = false
                    }
                }
                if (matchFound) {
                    // Match found; remove cards from dataset
                    gamePresenter.removeViewFlipper()
                    for ((_, card1) in facingUpCardsWeakHashMap) {
                        gamePresenter.notifyAdapterItemRemoved(card1.id)
                    }
                    facingUpCardsWeakHashMap.clear()
                    increaseGameScore()
                    gamePresenter.removeViewFlipper()
                } else {
                    // Match not found; turn cards over and start from the beginning.
                    decreaseGameScore()
                    gamePresenter.onTurnCardsOver()
                }
                gamePresenter.onGameScoreChanged(gameScore)
            }
        }
    }

    override fun removeCardsFromMap() {
        for ((key) in facingUpCardsWeakHashMap) {
            gamePresenter.notifyAdapterItemChanged(key)
        }
        facingUpCardsWeakHashMap.clear()
    }

    override fun shouldDispatchUiEvent(): Boolean {
        return facingUpCardsWeakHashMap.size != Constants.NUMBER_MATCHING_CARDS
    }

    override fun resetScore() {
        resetGameScore()
    }

    private fun shouldCheckForCardsMatch(): Boolean {
        return facingUpCardsWeakHashMap.size == Constants.NUMBER_MATCHING_CARDS
    }

    private fun increaseGameScore() {
        game.score = game.score + 1
    }

    private fun decreaseGameScore() {
        game.score = game.score - 1
    }

    private fun resetGameScore() {
        game.score = 0
    }
}
